waku/waku_archive/archive.nim

```diff
@@ -64,186 +75,148 @@ method validate*(validator: DefaultMessageValidator, msg: WakuMessage): Validati
+    hashes = query.hashes,
+    maxPageSize = maxPageSize + 1,
+    ascendingOrder = isAscendingOrder
+  )).valueOr:
+    return err(ArchiveError(kind: ArchiveErrorKind.DRIVER_ERROR, cause: error))
 
   let queryDuration = getTime().toUnixFloat() - queryStartTime
   waku_archive_query_duration_seconds.observe(queryDuration)
 
-  # Build response
-  if queryRes.isErr():
-    return err(ArchiveError(kind: ArchiveErrorKind.DRIVER_ERROR, cause: queryRes.error))
-
-  let rows = queryRes.get()
+  var hashes = newSeq[WakuMessageHash]()
   var messages = newSeq[WakuMessage]()
   var cursor = none(ArchiveCursor)
+  
   if rows.len == 0:
-    return ok(ArchiveResponse(messages: messages, cursor: cursor))
+    return ok(ArchiveResponse(hashes: hashes, messages: messages, cursor: cursor))
 
   ## Messages
-  let pageSize = min(rows.len, int(qMaxPageSize))
+  let pageSize = min(rows.len, int(maxPageSize))
+  
+  #TODO once store v2 is removed, unzip instead of 2x map
   messages = rows[0..<pageSize].mapIt(it[1])
+  hashes = rows[0..<pageSize].mapIt(it[4])
 
   ## Cursor
-  if rows.len > int(qMaxPageSize):
+  if rows.len > int(maxPageSize):
     ## Build last message cursor
     ## The cursor is built from the last message INCLUDED in the response
     ## (i.e. the second last message in the rows list)
-    let (pubsubTopic, message, digest, storeTimestamp) = rows[^2]
-
-    # TODO: Improve coherence of MessageDigest type
-    let messageDigest = block:
-        var data: array[32, byte]
-        for i in 0..<min(digest.len, 32):
-          data[i] = digest[i]
-
-        MessageDigest(data: data)
+    
+    #TODO Once Store v2 is removed keep only message and hash
+    let (pubsubTopic, message, digest, storeTimestamp, hash) = rows[^2]
 
+    #TODO Once Store v2 is removed, the cursor becomes the hash of the last message
     cursor = some(ArchiveCursor(
-      pubsubTopic: pubsubTopic,
-      senderTime: message.timestamp,
+      digest: MessageDigest.fromBytes(digest),
       storeTime: storeTimestamp,
-      digest: messageDigest
+      sendertime: message.timestamp,
+      pubsubTopic: pubsubTopic,
+      hash: hash,
     ))
 
   # All messages MUST be returned in chronological order
   if not isAscendingOrder:
     reverse(messages)
+    reverse(hashes)
 
-  return ok(ArchiveResponse(messages: messages, cursor: cursor))
+  return ok(ArchiveResponse(hashes: hashes, messages: messages, cursor: cursor))
 
-# Retention policy
-const WakuArchiveDefaultRetentionPolicyInterval* = chronos.minutes(30)
+proc periodicRetentionPolicy(self: WakuArchive) {.async.} =
+  debug "executing message retention policy"
 
-proc loopApplyRetentionPolicy*(w: WakuArchive):
-                               Future[Result[void, string]] {.async.} =
+  let policy = self.retentionPolicy.get()
 
-  if w.retentionPolicy.isNil():
-    return err("retentionPolicy is Nil in executeMessageRetentionPolicy")
+  while true:
+    await sleepAsync(WakuArchiveDefaultRetentionPolicyInterval)
```

Comment:
Would moving this sleep mean that the retention policy will execute only 30 mins after node startup? Is this the desired behaviour? To me it makes sense to run at startup and then every 30 mins.